File: api/reconcile.js
import ExcelJS from 'exceljs';

const MONTH_NAMES = ['','January','February','March','April','May','June','July',
  'August','September','October','November','December'];
const MONTH_TEXT = {
  1:['JAN','JANUARY'],2:['FEB','FEBRUARY'],3:['MAR','MARCH'],4:['APR','APRIL'],
  5:['MAY'],6:['JUNE','JUN'],7:['JULY','JUL'],8:['AUG','AUGUST'],
  9:['SEPT','SEP','SEPTEMBER'],10:['OCT','OCTOBER'],
  11:['NOV','NOVEMBER'],12:['DEC','DECEMBER']
};
const ACTUAL_LABELS   = ['ACTUAL ORDER SHIPPED','ACTUAL ORDER','ACTUAL SHIP'];
const FORECAST_LABELS = ['FORECAST'];
const ALL_LABELS = [...ACTUAL_LABELS,...FORECAST_LABELS,'BEG. IN STOCK','INCOMING','ON ORDER','IN STOCK'];

function extractAMG(text) {
  const m = String(text||'').match(/\b([A-Z]{2,4}\d{4,}(?:-CA\d|-TX)?)\b/);
  return m ? m[1] : null;
}

function extractNums(text) {
  const nums = String(text||'').match(/\b(\d{5,10})\b/g) || [];
  return nums.filter(n => !(n.length===4 && (n.startsWith('20')||n.startsWith('19'))));
}

function findMonthCol(rowVals, monthNum, yearNum) {
  const aliases = (MONTH_TEXT[monthNum]||[]).map(a=>a.toUpperCase());
  for (let ci=0; ci<rowVals.length; ci++) {
    const v = rowVals[ci];
    if (v == null) continue;
    if (typeof v === 'string') {
      const vu = v.trim().toUpperCase();
      if (aliases.some(a => vu===a || vu.startsWith(a))) return ci;
    }
    if (v instanceof Date) {
      if (v.getMonth()+1 === monthNum && v.getFullYear() === yearNum) return ci;
    }
  }
  return null;
}

async function loadSAP(buf, monthNum) {
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.load(buf);
  const ws = wb.worksheets[0];
  const headers = [];
  ws.getRow(1).eachCell((cell,ci) => { headers[ci-1] = String(cell.value||'').trim(); });
  const mname = MONTH_NAMES[monthNum].toLowerCase();
  const qColIdx = headers.findIndex(h => h.toLowerCase().includes(mname) && h.toLowerCase().includes('quantity'));
  if (qColIdx < 0) return [{}, `No quantity column for ${MONTH_NAMES[monthNum]}`];
  const data = {};
  ws.eachRow((row, ri) => {
    if (ri === 1) return;
    const itemNo = String(row.getCell(2).value||'').trim();
    if (!itemNo || itemNo==='null') return;
    const qty = row.getCell(qColIdx+1).value;
    if (qty != null) {
      const n = parseFloat(qty);
      if (!isNaN(n)) data[itemNo] = n;
    }
  });
  return [data, null];
}

async function process(sapBuf, fcBuf, tabName, monthNum, yearNum) {
  const [sapData, err] = await loadSAP(sapBuf, monthNum);
  if (err) return [null, err];

  const wb = new ExcelJS.Workbook();
  await wb.xlsx.load(fcBuf);

  const ws = wb.worksheets.find(s => s.name.toLowerCase().includes(tabName.toLowerCase().trim()));
  if (!ws) {
    const names = wb.worksheets.map(s=>s.name).join(', ');
    return [null, `Tab "${tabName}" not found. Available: ${names}`];
  }

  // Read all rows into array for easy processing
  const allRows = [];
  ws.eachRow((row) => {
    const vals = [];
    row.eachCell({includeEmpty:true}, (cell) => {
      vals.push(cell.value instanceof Object && cell.value.result !== undefined ? cell.value.result : cell.value);
    });
    allRows.push(vals);
  });

  const results = [];
  const updates = []; // {rowIdx (0-based), colIdx (0-based), value}

  let i = 0;
  while (i < allRows.length) {
    const row = allRows[i];
    if (!row || !row[0]) { i++; continue; }
    const cell0 = String(row[0]).trim();
    const amg = extractAMG(cell0);
    if (!amg || ALL_LABELS.some(l => cell0.toUpperCase().startsWith(l))) { i++; continue; }

    const mCol = findMonthCol(row, monthNum, yearNum);
    if (mCol === null) { i++; continue; }

    const itemNums = extractNums(cell0);
    let sapQty = null;
    for (const n of itemNums) { if (sapData[n] !== undefined) { sapQty = sapData[n]; break; } }
    if (sapQty === null && sapData[amg] !== undefined) sapQty = sapData[amg];

    let forecastVal = null, actualRowIdx = null;
    for (let j=i+1; j<Math.min(i+8, allRows.length); j++) {
      const r = allRows[j];
      if (!r || !r[0]) continue;
      const lbl = String(r[0]).trim().toUpperCase();
      if (extractAMG(String(r[0])) && !ALL_LABELS.some(l=>lbl.startsWith(l))) break;
      const raw = mCol < r.length ? r[mCol] : null;
      let val = null;
      if (raw != null && String(raw).trim() !== '' && String(raw).trim() !== 'CUT FROM PO') {
        const n = parseFloat(raw); if (!isNaN(n)) val = n;
      }
      if (FORECAST_LABELS.some(l=>lbl.startsWith(l)) && !lbl.includes('ACTUAL'))
        forecastVal = val ?? 0;
      else if (ACTUAL_LABELS.some(l=>lbl.startsWith(l)))
        actualRowIdx = j;
    }

    let status='no_sap', variance=null, variancePct=null;
    if (sapQty !== null) {
      variance = sapQty - (forecastVal||0);
      variancePct = forecastVal ? parseFloat(((variance/forecastVal)*100).toFixed(1)) : null;
      status = variance>0?'over':(variance<0?'under':'match');
    }

    results.push({amg, item_num:itemNums[0]||null, description:cell0.replace(/\n/g,' ').trim().substring(0,65),
      forecast:forecastVal, actual:sapQty, variance, variance_pct:variancePct, status,
      written: actualRowIdx!==null && sapQty!==null});

    if (actualRowIdx !== null && sapQty !== null)
      updates.push({rowIdx: actualRowIdx, colIdx: mCol, value: sapQty});

    i++;
  }

  // Apply updates — only touch the value, nothing else
  ws.eachRow((row, ri) => {
    const upd = updates.find(u => u.rowIdx === ri-1);
    if (upd) {
      const cell = row.getCell(upd.colIdx+1);
      cell.value = upd.value;
    }
  });

  const outBuf = await wb.xlsx.writeBuffer();
  const b64 = Buffer.from(outBuf).toString('base64');

  return [{
    file_b64: b64,
    sheet: ws.name,
    month: `${MONTH_NAMES[monthNum]} ${yearNum}`,
    items_found: results.length,
    written: updates.length,
    over:   results.filter(r=>r.status==='over').length,
    under:  results.filter(r=>r.status==='under').length,
    match:  results.filter(r=>r.status==='match').length,
    no_sap: results.filter(r=>r.status==='no_sap').length,
    results
  }, null];
}

export default async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  if (req.method === 'OPTIONS') { res.status(200).end(); return; }
  try {
    const { sap_b64, fc_b64, tab, month, year } = req.body;
    const sapBuf = Buffer.from(sap_b64, 'base64');
    const fcBuf  = Buffer.from(fc_b64,  'base64');
    const [result, err] = await process(sapBuf, fcBuf, tab, parseInt(month), parseInt(year));
    if (err) { res.status(400).json({ error: err }); return; }
    res.status(200).json(result);
  } catch(e) {
    res.status(500).json({ error: e.message });
  }
}

File: reconcile.py
import json, re, base64, io
from datetime import datetime
from openpyxl import load_workbook
 
MONTH_NAMES = ['','January','February','March','April','May','June','July',
               'August','September','October','November','December']
MONTH_TEXT = {
    1:['JAN','JANUARY'],2:['FEB','FEBRUARY'],3:['MAR','MARCH'],4:['APR','APRIL'],
    5:['MAY'],6:['JUNE','JUN'],7:['JULY','JUL'],8:['AUG','AUGUST'],
    9:['SEPT','SEP','SEPTEMBER'],10:['OCT','OCTOBER'],
    11:['NOV','NOVEMBER'],12:['DEC','DECEMBER']
}
ACTUAL_LABELS   = ['ACTUAL ORDER SHIPPED','ACTUAL ORDER','ACTUAL SHIP']
FORECAST_LABELS = ['FORECAST']
ALL_LABELS = ACTUAL_LABELS + FORECAST_LABELS + ['BEG. IN STOCK','INCOMING','ON ORDER','IN STOCK']
 
def extract_amg(text):
    m = re.search(r'\b([A-Z]{2,4}\d{4,}(?:-CA\d|-TX)?)\b', str(text))
    return m.group(1) if m else None
 
def extract_nums(text):
    nums = re.findall(r'\b(\d{5,10})\b', str(text))
    return [n for n in nums if not (len(n)==4 and n[:2] in ('20','19'))]
 
def find_month_col(row_vals, month_num, year_num):
    aliases = [a.upper() for a in MONTH_TEXT.get(month_num, [])]
    for ci, v in enumerate(row_vals):
        if v is None: continue
        if isinstance(v, str):
            vu = v.strip().upper()
            if any(vu == a or vu.startswith(a) for a in aliases): return ci
        if isinstance(v, datetime):
            if v.month == month_num and v.year == year_num: return ci
    return None
 
def load_sap(data_bytes, month_num):
    wb = load_workbook(io.BytesIO(data_bytes), read_only=True, data_only=True)
    ws = wb.active
    rows = list(ws.iter_rows(values_only=True))
    headers = [str(c).strip() if c else '' for c in rows[0]]
    mname = MONTH_NAMES[month_num].lower()
    qty_col = next((i for i,h in enumerate(headers) if mname in h.lower() and 'quantity' in h.lower()), None)
    if qty_col is None:
        wb.close(); return {}, f"No quantity column for {MONTH_NAMES[month_num]}"
    data = {}
    for row in rows[1:]:
        item_no = str(row[1]).strip() if row[1] else None
        if item_no and item_no != 'None':
            qty = row[qty_col]
            if qty is not None:
                try: data[item_no] = float(qty)
                except: pass
    wb.close()
    return data, None
 
def process(sap_bytes, fc_bytes, tab_name, month_num, year_num):
    sap_data, err = load_sap(sap_bytes, month_num)
    if err: return None, err
 
    wb = load_workbook(io.BytesIO(fc_bytes), data_only=False, keep_vba=False)
    sheet_name = next((n for n in wb.sheetnames if tab_name.lower().strip() in n.lower().strip()), None)
    if not sheet_name:
        return None, f'Tab "{tab_name}" not found. Available: {", ".join(wb.sheetnames)}'
 
    ws = wb[sheet_name]
    wb_ro = load_workbook(io.BytesIO(fc_bytes), read_only=True, data_only=True)
    ws_ro = wb_ro[sheet_name]
    all_rows = list(ws_ro.iter_rows(values_only=True))
    wb_ro.close()
 
    results, updates = [], []
    i = 0
    while i < len(all_rows):
        row = all_rows[i]
        if not row or not row[0]: i += 1; continue
        cell0 = str(row[0]).strip()
        amg = extract_amg(cell0)
        if not amg or any(cell0.upper().startswith(l) for l in ALL_LABELS): i += 1; continue
        month_col = find_month_col(row, month_num, year_num)
        if month_col is None: i += 1; continue
        item_nums = extract_nums(cell0)
        sap_qty = None
        for num in item_nums:
            if num in sap_data: sap_qty = sap_data[num]; break
        if sap_qty is None and amg in sap_data: sap_qty = sap_data[amg]
        forecast_val = None
        actual_row_idx = None
        for j in range(i+1, min(i+8, len(all_rows))):
            r = all_rows[j]
            if not r or not r[0]: continue
            lbl = str(r[0]).strip().upper()
            if extract_amg(str(r[0])) and not any(lbl.startswith(l) for l in ALL_LABELS): break
            raw = r[month_col] if len(r) > month_col else None
            try: val = float(raw) if raw is not None and str(raw).strip() not in ('','None','CUT FROM PO') else None
            except: val = None
            if any(lbl.startswith(l) for l in FORECAST_LABELS) and 'ACTUAL' not in lbl:
                forecast_val = val if val is not None else 0
            elif any(lbl.startswith(l) for l in ACTUAL_LABELS):
                actual_row_idx = j
        status = 'no_sap'
        variance = variance_pct = None
        if sap_qty is not None:
            variance = sap_qty - (forecast_val or 0)
            variance_pct = round((variance/forecast_val*100),1) if forecast_val else None
            status = 'over' if variance > 0 else ('under' if variance < 0 else 'match')
        results.append({'amg':amg,'item_num':item_nums[0] if item_nums else None,'description':cell0.replace('\n',' ').strip()[:65],'forecast':forecast_val,'actual':sap_qty,'variance':variance,'variance_pct':variance_pct,'status':status,'written':actual_row_idx is not None and sap_qty is not None})
        if actual_row_idx is not None and sap_qty is not None:
            updates.append((actual_row_idx+1, month_col+1, sap_qty))
        i += 1
 
    for (excel_row, excel_col, val) in updates:
        ws.cell(row=excel_row, column=excel_col).value = val
 
    out = io.BytesIO()
    wb.save(out); out.seek(0)
    return {'file_b64':base64.b64encode(out.read()).decode(),'sheet':sheet_name,'month':f'{MONTH_NAMES[month_num]} {year_num}','items_found':len(results),'written':len(updates),'over':sum(1 for r in results if r['status']=='over'),'under':sum(1 for r in results if r['status']=='under'),'match':sum(1 for r in results if r['status']=='match'),'no_sap':sum(1 for r in results if r['status']=='no_sap'),'results':results}, None
 
 
def app(environ, start_response):
    method = environ.get('REQUEST_METHOD','')
    cors = [('Access-Control-Allow-Origin','*'),('Access-Control-Allow-Methods','POST, OPTIONS'),('Access-Control-Allow-Headers','Content-Type'),('Content-Type','application/json')]
    if method == 'OPTIONS':
        start_response('200 OK', cors); return [b'']
    try:
        length = int(environ.get('CONTENT_LENGTH',0))
        body = environ['wsgi.input'].read(length)
        payload = json.loads(body)
        sap_bytes = base64.b64decode(payload['sap_b64'])
        fc_bytes  = base64.b64decode(payload['fc_b64'])
        result, err = process(sap_bytes, fc_bytes, payload['tab'], int(payload['month']), int(payload['year']))
        if err:
            start_response('400 Bad Request', cors)
            return [json.dumps({'error':err}).encode()]
        start_response('200 OK', cors)
        return [json.dumps(result).encode()]
    except Exception as e:
        start_response('500 Internal Server Error', cors)
        return [json.dumps({'error':str(e)}).encode()]
